soccer/optimization/ParallelGradientAscent1D.cpp

```diff
@@ -0,0 +1,74 @@
+#include "ParallelGradientAscent1D.hpp"
+#include <math.h>
+#include <iostream>
+
+ParallelGradientAscent1D::ParallelGradientAscent1D(
+    ParallelGradient1DConfig* config)
+    : config(config) {
+    // Create list of GA1Ds
+    for (int i = 0; i < (config->GA1DConfig).size(); i++) {
+        GA1Ds.push_back(GradientAscent1D(&config->GA1DConfig.at(i)));
+    }
+}
+
+/**
+ * Executes all GA1Ds until their max has been reached
+ */
+void ParallelGradientAscent1D::execute() {
+    // While any are not done
+    bool continueExecution = true;
+
+    while (continueExecution) {
+        // Default to false unless any still need to work
+        continueExecution = false;
+
+        // Execute a step for each one
+        for (auto& GA1D : GA1Ds) {
+            if (GA1D.continueExecution()) {
+                GA1D.singleStep();
+                continueExecution = true;
+            }
+        }
+
+        // Assume ascending order for xStart
+        for (int i = 0; i < GA1Ds.size() - 1; i++) {
+            GradientAscent1D lower = GA1Ds.at(i);
+            GradientAscent1D upper = GA1Ds.at(i + 1);
+
+            // Erase elements if they get too close
+            // This helps kill any GA1Ds that are going up the same hill
+            if (fabs(lower.getXValue() - upper.getXValue()) <
+                config->xCombineThresh) {
+                GA1Ds.erase(GA1Ds.begin() + i + 1);
```

Comment:
Oh right the off by one error at the end. Hmm, yeah not really worth it i guess unless it was actually speeding it up. You'd have to iterate to the end -1 instead, then copy the last one before you erase everything after. Something like.
int i = 1;
auto end = GA1Ds.end()-1;
auto last = std::remove_if(GA1Ds.begin(), GA1Ds.end()-1, [this, &i](auto& value) {
                auto& next = GA1Ds[i];
                i++;
                return fabs(value.getXValue() - next.getXValue()) <
                       config->xCombineThresh;
            });
*last = *end;
GA1Ds.erase(last + 1);